src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs

```diff
@@ -307,128 +402,144 @@ public static (uint encoded, int bitLength) Encode(int data)
-                // from src[i]. Thus we need to read 5 bytes here to ensure that we always have
-                // at least 30 bits available for decoding.
-                // TODO https://github.com/dotnet/runtime/issues/1506:
-                // Rework this as part of Huffman perf improvements
-                uint next = (uint)(src[i] << 24 + lastDecodedBits);
-                next |= (i + 1 < src.Length ? (uint)(src[i + 1] << 16 + lastDecodedBits) : 0);
-                next |= (i + 2 < src.Length ? (uint)(src[i + 2] << 8 + lastDecodedBits) : 0);
-                next |= (i + 3 < src.Length ? (uint)(src[i + 3] << lastDecodedBits) : 0);
-                next |= (i + 4 < src.Length ? (uint)(src[i + 4] >> (8 - lastDecodedBits)) : 0);
-
-                uint ones = (uint)(int.MinValue >> (8 - lastDecodedBits - 1));
-                if (i == src.Length - 1 && lastDecodedBits > 0 && (next & ones) == ones)
-                {
-                    // The remaining 7 or less bits are all 1, which is padding.
-                    // We specifically check that lastDecodedBits > 0 because padding
-                    // longer than 7 bits should be treated as a decoding error.
-                    // http://httpwg.org/specs/rfc7541.html#rfc.section.5.2
-                    break;
-                }
-
-                // The longest possible symbol size is 30 bits. If we're at the last 4 bytes
-                // of the input, we need to make sure we pass the correct number of valid bits
-                // left, otherwise the trailing 0s in next may form a valid symbol.
-                int validBits = Math.Min(30, (8 - lastDecodedBits) + (src.Length - i - 1) * 8);
-                int ch = DecodeValue(next, validBits, out int decodedBits);
-
-                if (ch == -1)
-                {
-                    // No valid symbol could be decoded with the bits in next
-                    throw new HuffmanDecodingException(SR.net_http_hpack_huffman_decode_failed);
-                }
-                else if (ch == 256)
-                {
-                    // A Huffman-encoded string literal containing the EOS symbol MUST be treated as a decoding error.
-                    // http://httpwg.org/specs/rfc7541.html#rfc.section.5.2
-                    throw new HuffmanDecodingException(SR.net_http_hpack_huffman_decode_failed);
-                }
+                // Load next 8 bits into accumulator.
+                acc <<= 8;
+                acc |= src[i++];
+                bitsInAcc += 8;
 
-                if (j == dst.Length)
+                // Decode bits in accumulator.
+                do
                 {
-                    Array.Resize(ref dstArray, dst.Length * 2);
-                    dst = dstArray;
-                }
+                    lookupIndex = (byte)(acc >> (bitsInAcc - 8));
 
-                dst[j++] = (byte)ch;
+                    ushort lookupValue = decodingTree[(lookupTableIndex << 8) + lookupIndex];
```

Comment:
```suggestion
                    uint lookupValue = decodingTree[(lookupTableIndex << 8) + lookupIndex];
```
Same reason, avoid partial registers, as they need to be zero-extended (`movzx`).

Reply:
done